### ISSUE TITLE
Access to each batch's gradients while accumulating gradients?

### ISSUE THREAD

**Issue by JJumSSu** [pytorch/fairseq#1712]
Hello, 

I have one question regarding the gradient accumulation.
Fairseq provides "gradient accumulation", which accumulates gradients for several batches.

From my understanding, the loss.backward() stores the gradients in variables' ```.grad``` which gets summed up for every 8 batches(If accumulating gradients for 8 batches). And optimizer.step() (which is called every update-frequency, e.g. 8) will use these "summed" gradients for the update of the parameters.

However, I'd like to get access to each batch's gradients(after every loss.backward()) while training with gradient accumulation. But as far as I know, the gradients stored in ```.grad``` are summed up, so gradients from other bathes are mixed up.

Is there an easy way of getting each batch's gradient information?

Thank you!

**Comment by myleott** [comment]
Hmm, one option is to disable fairseq's gradient accumulation, then define an optimizer that does the accumulation internally (and in the process gives you access to the intermediate gradients).

**Comment by JJumSSu** [comment]
@myleott 
Thank you for your reply!

Can you explain about the optimizer that does the accumulation internally? From my understanding, after calling loss.backward() for several batches, the gradients are accumulated(summed) in ```.grad``` variable. And optimizer steps with gradients information stored in ```.grad```. If I turn off the gradient accumulation, then because the zero.grad() gets called every batch,

**Comment by myleott** [comment]
In the optimizer you could loop over the params, copy/accumulate the gradients into another buffer, and every k steps update the model params. Something like this (I haven't tested it though):

```python
for group in self.param_groups:
  for p in group['params']:
    if p.grad is None:
      continue

    # accumulate grads
    if p in self.acc_grads:
      self.acc_grads[p] += p.grad
    else:
      self.acc_grads[p] = p.grad.clone()
    self.num_acc += 1

    if self.num_acc % num_to_accumulate == 0
      # do regular step, using acc_grads[p]

      # reset accumulation
      self.num_acc = 0
      self.acc_grads = {}
```

**Comment by JJumSSu** [comment]
@myleott 
Thank you for such a nice and meticulous reply.

**Comment by lematt1991** [comment]
Seems this has been resolved.  Please open a new issue if you are still having problems.
